### PR TITLE
Possible option: use `ParamSpec` for `lru_cache`

### PR DESCRIPTION
**Pull request by sobolevn**
This one might be controversal. Why?
- When adding `ParamSpec` we cannot ensure that args are `Hashable`, but signature remains the same
- When not using `ParamSpec`, we can ensure that args are `Hashable`, but signature changes to `*Any, **Any`

What do others think? In my opinion better signature is more important 🤔 

Local testing with the latest mypy shows that my version works fine:

```python
from typing import TypeVar, Callable, ParamSpec, Generic

_T = TypeVar("_T")
_P = ParamSpec("_P")

class _lru_cache_wrapper(Generic[_P, _T]):
    __wrapped__: Callable[_P, _T]
    def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _T: ...

def lru_cache(maxsize: Callable[_P, _T]) -> _lru_cache_wrapper[_P, _T]:
    pass

@lru_cache
def some(arg: int) -> str:
    pass

reveal_type(some)  # N: Revealed type is "ex._lru_cache_wrapper[def (arg: builtins.int), builtins.str*]"
```

### CODE REVIEW

**Comment by sobolevn** [comment]
This looks like a mypy bug:

```
 stdlib/functools.pyi:27: error: Argument 1 to "final" has incompatible type "Type[_lru_cache_wrapper[Any, Any]]"; expected "Type[_lru_cache_wrapper[Any, Any]]"  [arg-type]
```

**Comment by sobolevn** [comment]
Looks like we have another problem, here's an example from `flake8`:

```python
@functools.lru_cache(maxsize=None)
def style_guide_for(self, filename: str) -> "StyleGuide": pass
```

It is then used: `guide = self.style_guide_for(filename)`
So, it report:

```
+ src/flake8/style_guide.py:275: error: Missing positional argument "filename" in call to "__call__" of "_lru_cache_wrapper"
+ src/flake8/style_guide.py:275: error: Argument 1 to "__call__" of "_lru_cache_wrapper" has incompatible type "str"; expected "StyleGuideManager"
```

I guess that `self` arg breaks it 😒

**Comment by AlexWaygood** [comment]
We had a looong discussion about how to type `lru_cache` over in #6347, but didn't really come to a conclusion ://

**Comment by sobolevn** [comment]
Going to try `@staticmethod` approach!

File: stdlib/functools.pyi [hunk 1]
```diff
@@ -24,20 +25,21 @@ class _CacheInfo(NamedTuple):
...
     def cache_info(self) -> _CacheInfo: ...
     def cache_clear(self) -> None: ...
 
 if sys.version_info >= (3, 8):
     @overload
-    def lru_cache(maxsize: int | None = ..., typed: bool = ...) -> Callable[[Callable[..., _T]], _lru_cache_wrapper[_T]]: ...
+    def lru_cache(maxsize: int | None = ..., typed: bool = ...) -> Callable[[Callable[_P, _T]], _lru_cache_wrapper[_P, _T]]: ...
     @overload
-    def lru_cache(maxsize: Callable[..., _T], typed: bool = ...) -> _lru_cache_wrapper[_T]: ...
+    def lru_cache(maxsize: Callable[_P, _T]) -> _lru_cache_wrapper[_P, _T]: ...
```
Comment (AlexWaygood) [inline stdlib/functools.pyi:39]: Why'd the `typed` arg get deleted? :)

Comment (sobolevn) [reply stdlib/functools.pyi:39]: Yeap, my bad. It is allowed: https://github.com/python/cpython/blob/54610bb448a9cf5be77d53b66169fca4c11be6cb/Lib/functools.py#L508

**Comment by sobolevn** [comment]
Ok, it didn't work out. I am going to close it and at least way for https://github.com/python/mypy/issues/12033 to be merged and `Concatenate` to be supported.

**Comment by tamird** [comment]
@sobolevn is it time to revisit this?

**Comment by sobolevn** [comment]
@tamird please, go ahead and try!

**Comment by tamird** [comment]
I'm out of my depth on trying, just another frustrated user suffering bad typings :(

**Comment by sobolevn** [comment]
Looking at some recent discussion, this seems to be rather hard / impossible right now: https://github.com/python/typeshed/pull/10523

**Pull request closed without merging**